### ISSUE TITLE
uint instead of u32 for sizes?

### ISSUE THREAD

**Issue by flying-sheep** [PistonDevelopers/image#230]
`uint` is basically `size_t` for rust, so wouldn’t it make sense to use it for x, y, width, height, …?

this would change GenericImage to

```rust
impl GenericImage<Rgba<u8>> for DynamicImage
    fn dimensions(&self) -> (uint, uint)
    fn bounds(&self) -> (uint, uint, uint, uint)
    fn get_pixel(&self, x: uint, y: uint) -> Rgba<u8>
    fn put_pixel(&mut self, x: uint, y: uint, pixel: Rgba<u8>)
    fn blend_pixel(&mut self, x: uint, y: uint, pixel: Rgba<u8>)
    fn get_pixel_mut(&mut self, _: uint, _: uint) -> &mut Rgba<u8>
    fn pixels(&self) -> Pixels<Self>
    fn pixels_mut(&mut self) -> MutPixels<Self>
```

**Comment by sivadeilra** [comment]
I think this is worse, not better.  The reasonable range of sizes for an image has nothing to do with the pointer size on a particular machine.

**Comment by nwin** [comment]
No, using `uint` for other things than pointer addresses is rarely a good idea. We might consider using u64 for that, but our library is not prepared to be used for excessively big images that would never fit into the RAM.

**Comment by eddyb** [comment]
AFAIK PNG limits width and height to 32-bit so anything larger than that wouldn't be encodable as a PNG.
Even if other formats didn't have similar restrictions (which I think they do), it would be a bad idea to support image sized not supported by your primary lossless file format.

**Comment by flying-sheep** [comment]
my pain comes from the fact that [japaric/linalg.rs](https://github.com/japaric/linalg.rs) uses uints, leading to this kind of code:

```rust
let (w, h) = img.dimensions();
let dims = (
	w.to_uint().unwrap(),
	h.to_uint().unwrap(),
);

Mat::from_fn(dims, |(x, y)| rgba_to_labt(img.get_pixel(x as u32, y as u32))).unwrap()
```

i was under the impression that we shouldn’t use pointer-sized datatypes for *data*, as in “things we calculate with”, but use u32 and friends for that, unless we need 64 bits of size/precision (because compilers make 32 bit math faster than 64 bit processor word sized math).

which still makes uint the type of choice for “references that you don’t care about the size of”

**Comment by nwin** [comment]
Well, we care about the size because it is defined as u32 in all image format we support so far. Furthermore the dimensions of an image are data.

**Comment by flying-sheep** [comment]
i view coordinates as references. So you think the API of japaric/linalg should be changed to u32s?

**Comment by nwin** [comment]
No. That represents arbitrary vectors/matrices, not images that have a well-defined range of values for their coordinates.

**Comment by flying-sheep** [comment]
> images that have a well-defined range of values for their coordinates

i doubt that dense matrices with a dimension of more than 2^32 are more practical than pictures with this dimension

**Comment by bvssvni** [comment]
It's not a problem to release a 2.0 version with support for larger dimensions, if needed.
